### PR TITLE
added support for migration without crq

### LINKED ISSUE #7: add support for migration of subnamespaces whose direct parents do not have a CRQ
At the moment it's only allowed to perform migration on subnamespaces whose parent has a CRQ.

We should add support such that migration is possible even in these levels of the hierarchy.

/good-first-issue

### PR DESCRIPTION
**Pull request by mzeevi**
Fixes #7. With this PR it is now possible to do a migration hierarchy in every level of the hierarchy. In addition, fixes were made to the e2e tests

### CODE REVIEW

File: internals/utils/quota_utils.go [hunk 1]
```diff
@@ -185,19 +259,35 @@ func DoesSNSQuotaObjectExist(sns *ObjectContext) (bool, *ObjectContext, error) {
 	return false, nil, nil
 }
 
-// DoesSNSCrqExists returns true if a ClusterResourceQuota exists
-func DoesSNSCrqExists(sns *ObjectContext) bool {
-	snsCrq, err := NewObjectContext(sns.Ctx, sns.Client, types.NamespacedName{Name: sns.Object.GetName()}, &quotav1.ClusterResourceQuota{})
+// DoesSNSCRQExists returns true if a ClusterResourceQuota exists
+func DoesSNSCRQExists(sns *ObjectContext) (bool, *ObjectContext, error) {
```
Comment (omerbd21) [inline internals/utils/quota_utils.go:263]: why use a bool and not just check if snsCrq != nil?

File: internals/utils/quota_utils.go [hunk 2]
```diff
@@ -185,19 +259,35 @@ func DoesSNSQuotaObjectExist(sns *ObjectContext) (bool, *ObjectContext, error) {
...
+			return true, snsCrq, nil
 		}
 	}
-	return false
+
+	return false, nil, nil
+}
+
+// DoesSNSRQExists returns true if a ResourceQuota exists
+func DoesSNSRQExists(sns *ObjectContext) (bool, *ObjectContext, error) {
```
Comment (omerbd21) [inline internals/utils/quota_utils.go:279]: same comment as in DoesSNSCRQExists

**Review by omerbd21 — CHANGES_REQUESTED** [review]

File: internals/controllers/migrationhierarchy/migrationhierarchy_controller.go [hunk 3]
```diff
@@ -108,8 +108,22 @@ func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext)
...
 	if phase == danav1.None {
 		if sourceQuotaObjExists {
-			if err := createMigrationUPQ(mhObject, sourceResources, rootNSName, toNamespace); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed create updateQuota for migration '%s': "+err.Error(), mhObject.GetName())
+			// temporarily increase the quota of the root namespace to make
+			// sure the migrationUPQ doesn't result in an error due to insufficent resources
+			if er := increaseRootResources(mhObject, rootNSName, sourceResources); err != nil {
+				err := r.updateMHStatus(mhObject, danav1.Error, er.Error())
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/migrationhierarchy/migrationhierarchy_controller.go:116]: It's consistent like this in the entire codebase so I think it's fine like this

File: internals/controllers/migrationhierarchy/migrationhierarchy_controller.go [hunk 4]
```diff
@@ -195,6 +209,11 @@ func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext)
 		if err := createMigrationUPQ(mhObject, sourceResources, sourceSNSParentName, rootNSName); err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to create updateQuota for migration '%s': "+err.Error(), mhObject.GetName())
 		}
+
+		// Decrease the quota of the root namespace which was temporarily added
+		if err := decreaseRootResources(mhObject, rootNSName, sourceResources); err != nil {
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/migrationhierarchy/migrationhierarchy_controller.go:214]: Good call

File: internals/controllers/updatequota/updatequota_controller.go [hunk 5]
```diff
@@ -281,19 +281,19 @@ func addSnsQuota(sns *utils.ObjectContext, quotaSpec corev1.ResourceQuotaSpec) e
...
 // loops through each resource, subtracting the requested quota if it is specified
 func subSnsQuota(sns *utils.ObjectContext, quotaSpec corev1.ResourceQuotaSpec) error {
 	err := sns.EnsureUpdateObject(func(object client.Object, l logr.Logger) (client.Object, logr.Logger, error) {
 		snsQuota := object.(*danav1.Subnamespace).Spec.ResourceQuotaSpec
-		for res := range snsQuota.Hard {
+		for resourceName := range snsQuota.Hard {
 			var (
-				vBefore, _  = snsQuota.Hard[res]
-				vRequest, _ = quotaSpec.Hard[res]
+				vBefore, _  = snsQuota.Hard[resourceName]
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/updatequota/updatequota_controller.go:292]: Changed

File: internals/controllers/subnamespace/subnamespace_sync.go [hunk 6]
```diff
@@ -67,10 +66,21 @@ func (r *SubnamespaceReconciler) sync(snsParentNS, snsObject *utils.ObjectContex
 	// if the subnamespace is a regular SNS (i.e. not a ResourcePool) OR it's an upper-rp, then create a corresponding
 	// quota object for the subnamespace. The quota object can be either a ResourceQuota or a ClusterResourceQuota
 	if !isSNSResourcePool || isSNSUpperResourcePool {
-		if rqFlag {
-			err := deleteLegacyCRQ(snsObject)
-			if err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to delete legacy quota object for subnamespace '%s': "+err.Error(), snsName)
+		exists, quotaObject, err := utils.DoesSNSCRQExists(snsObject)
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/subnamespace/subnamespace_sync.go:69]: This handles a situation when a subnamespace used to have a CRQ but should have a RQ, and vice-versa. It's relevant in migrations

File: internals/utils/quota_utils.go [hunk 7]
```diff
@@ -62,37 +63,82 @@ func GetSNSQuotaObject(sns *ObjectContext) (*ObjectContext, error) {
...
+	quotaObject, err := NewObjectContext(sns.Ctx, sns.Client, client.ObjectKey{Namespace: "", Name: sns.Object.GetName()}, &quotav1.ClusterResourceQuota{})
+	if err != nil {
+		return quotaObject, err
+	}
+
+	return quotaObject, nil
+}
+
+// DeleteSNSRQ ensures deletion of the RQ quota object
+func DeleteSNSRQ(sns *ObjectContext) error {
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/utils/quota_utils.go:104]: I don't actually use neither function and just deleted them instead

File: internals/utils/quota_utils.go (hunk 1 above)
Comment (mzeevi) [reply internals/utils/quota_utils.go:263]: It's more consistent with how these checks are done across the code (not saying it's the best way)

Comment (omerbd21) [reply internals/utils/quota_utils.go:233]: I think it's creating unnecessary variables and we should consider changing it across the code. for now i'll approve.

**Review by omerbd21 — APPROVED** [review]

**Pull request merged**